File: 2023/day1/ts/main.ts
const file = Deno.readTextFileSync('../input.txt')
const lines = file.split('\n').filter((line) => line.trim() !== '')

interface Match {
  index: number
  digit: string
  type: 'digit' | 'word'
}

const lookup: { [key: string]: string } = {
  one: '1',
  two: '2',
  three: '3',
  four: '4',
  five: '5',
  six: '6',
  seven: '7',
  eight: '8',
  nine: '9',
}

let sum = 0

for (const line of lines) {
  const digits: Match[] = []

  // find all digits in the line
  for (let i = 0; i < line.length; i++) {
    const char = line[i]
    if (char >= '0' && char <= '9') {
      digits.push({ index: i, digit: char, type: 'digit' })
    }
  }

  // find first and last unique spelled numbers in the line
  for (const key in lookup) {
    const firstMatch = line.indexOf(key)
    if (firstMatch !== -1) {
      digits.push({ index: firstMatch, digit: lookup[key], type: 'word' })
    }

    const lastMatch = line.lastIndexOf(key)
    if (lastMatch !== -1 && lastMatch !== firstMatch) {
      digits.push({ index: lastMatch, digit: lookup[key], type: 'word' })
    }
  }

  // sort digits by index
  digits.sort((a, b) => a.index - b.index)

  // concat first and last digits
  const num = Number(digits[0].digit + digits[digits.length - 1].digit)

  sum += num
}

console.log(sum)

File: 2023/day1/golang/main.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func main() {
	content, err := os.ReadFile("../input.txt")
	if err != nil {
		panic(err)
	}
	lines := strings.Split(string(content), "\n")

	// strip empty lines
	lines = slices.DeleteFunc(lines, func(line string) bool {
		return line == ""
	})

	// part 1
	sumPart1 := 0
	for _, line := range lines {
		var digits []string

		// Check each position in the string for digits
		for i := 0; i < len(line); i++ {
			if line[i] >= '0' && line[i] <= '9' {
				digits = append(digits, string(line[i]))
			}
		}

		if len(digits) == 0 {
			continue
		}

		// Take first and last digit
		first := digits[0]
		last := digits[len(digits)-1]

		num, err := strconv.Atoi(first + last)
		if err != nil {
			panic(err)
		}
		sumPart1 += num
	}

	// part 2
	stringToNumber := map[string]string{
		"one":   "1",
		"two":   "2",
		"three": "3",
		"four":  "4",
		"five":  "5",
		"six":   "6",
		"seven": "7",
		"eight": "8",
		"nine":  "9",
	}

	sumPart2 := 0
	for _, line := range lines {
		var digits []string

		// Check each position in the string for either a digit or a word
		for i := 0; i < len(line); i++ {
			// Check for numeric digit
			if line[i] >= '0' && line[i] <= '9' {
				digits = append(digits, string(line[i]))
				continue
			}

			// Check for word numbers
			for word, num := range stringToNumber {
				if strings.HasPrefix(line[i:], word) {
					digits = append(digits, num)
				}
			}
		}

		if len(digits) == 0 {
			continue
		}

		// Take first and last digit
		first := digits[0]
		last := digits[len(digits)-1]

		num, err := strconv.Atoi(first + last)
		if err != nil {
			panic(err)
		}
		sumPart2 += num
	}

	fmt.Println(sumPart1)
	fmt.Println(sumPart2)
}
